book/src/development/emitting_lints.md

```diff
@@ -0,0 +1,217 @@
+
+```rust
+pub trait LateLintPass<'tcx>: LintPass {
+    // Trait methods
+}
+```
+
+By far the most common method used for Clippy lints is [`check_expr` method][late_check_expr],
+this is likely because Rust is an expression language and, more often than not,
+the lint we want to work on must examine expressions.
+
+> _Note:_ If you don't fully understand what expressions are in Rust,
+> take a look at the official documentation on [expressions][rust_expressions]
+
+Other common ones include the [`check_fn` method][late_check_fn] and the
+[`check_item` method][late_check_item]. We choose to implement whichever trait
+method based on what we need for the lint at hand.
+
+### Implement Trait Method
+
+Assume that we have added and defined a `bar_expressions` lint, we could write
+down a skeleton for this lint as the following:
+
+```rust
+impl<'tcx> LateLintPass<'tcx> for BarExpressions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>)  {}
+}
+```
+
+### Emitting a lint
+
+Inside the trait method that we implement, we can write down the lint logic
+and the emit the lint with suggestions.
+
+Clippy's [diagnostics] provides quite a few diagnostic functions that we can
+use to emit lints. Take a look at the documentation to pick one that suits
+your lint's needs the best. Some common ones you will encounter in the Clippy
+repository includes:
+
+- [span_lint_and_help]: emits a lint and provides a helpful message
+- [span_lint_and_sugg]: emits a lint and provides a suggestion to fix the code
+
+```rust
+impl<'tcx> LateLintPass<'tcx> for BarExpressions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>)  {
+        // Imagine that `some_bar_expr_logic` checks for requirements for emitting the lint
+        if some_bar_expr_logic(expr) {
+            span_lint_and_help(
+                cx, // < The context
+                FOO_FUNCTIONS, // < The name of the lint in ALL CAPS
+                expr.span, // < The span to lint
+                "message on why the lint is emitted",
+                None, // < An optional help span (to highlight something in the lint)
+                "message that provides a helpful suggestion",
+            );
+        }
+    }
+}
+```
+
+> Note: According to [the rustc-dev-guide], the message should be matter of fact and avoid
+> capitalization and periods, unless multiple sentences are needed. When code or
+> an identifier must appear in a message or label, it should be surrounded with
+> single grave accents (\`).
+
+## Suggestions: Automatic fixes
+
+Some lints know what to change in order to fix the lint. A real world example, the lint
+[`range_plus`][range_plus_one] lints for ranges where the user wrote `x..y + 1` instead of using an
+[inclusive range][inclusive_range] (`x..=1`). The fix to this lint would be just changing your `x..y + 1` expression
+to `x..=y`, **this is where suggestions come in**.
```

Comment:
This paragraph is worded weirdly.  also the suggestion doesn't fix the lint, but the code.